src/dart/lib/widgets.dart

```diff
@@ -0,0 +1,319 @@
+      _bytes = null;
+      _fetchBytes();
+    }
+  }
+
+  @override
+  void didUpdateWidget(IconVGImage oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.asset != widget.asset) {
+      _bytes = null;
+      _fetchBytes();
+    }
+  }
+
+  Future<void> _fetchBytes() async {
+    assert(_assetBundle != null);
+    final ByteData data = await _assetBundle!.load(widget.asset);
+    if (mounted) {
+      setState(() {
+        _bytes = data.buffer.asUint8List();
+      });
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return RawIconVGImage(
+      bytes: _bytes,
+      palette: widget.palette,
+      size: widget.size,
+      clipBehavior: widget.clipBehavior,
+      semanticLabel: widget.semanticLabel,
+    );
+  }
+}
+
+/// Render the specified [bytes] as an IconVG image.
+///
+/// If the image is an asset, consider using [IconVGImage] instead.
+class RawIconVGImage extends StatefulWidget {
+  const RawIconVGImage({
+    Key? key,
+    this.bytes,
+    this.palette,
+    this.size,
+    this.clipBehavior = Clip.hardEdge,
+    this.semanticLabel,
+  }) : super(key: key);
+
+  /// The bytes. Must be in the IconVG format.
+  final Uint8List? bytes;
+
+  /// The custom color palette used to decode the image.
+  ///
+  /// Must not have more than 64 colors.
+  final List<Color>? palette;
+
+  /// The target render size. Defaults to the [BoxConstraints.maxHeight].
+  final Size? size;
+
+  /// How to clip the image.
+  ///
+  /// This can be used to enable anti-aliased clips if the image is rotated, or
+  /// disable clips entirely (for improved performance) if the image does not
+  /// need them.
+  final Clip clipBehavior;
+
+  /// A Semantic description of the image.
+  ///
+  /// Used to provide a description of the image to TalkBack on Android, and
+  /// VoiceOver on iOS.
+  final String? semanticLabel;
+
+  @override
+  State<RawIconVGImage> createState() => _RawIconVGImageState();
+}
+
+class _RawIconVGImageState extends State<RawIconVGImage> {
+  IconVGFile? _image;
+  double? _decodedHeight;
+
+  @override
+  void didUpdateWidget(RawIconVGImage oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.bytes != widget.bytes ||
+        oldWidget.palette != widget.palette ||
+        oldWidget.clipBehavior != widget.clipBehavior) {
+        // widget.size is handled in build
+      _image = null;
+    }
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (widget.bytes == null) {
+      return SizedBox.fromSize(size: widget.size ?? Size.zero);
+    }
+    return Semantics(
+      container: widget.semanticLabel != null,
+      image: true,
+      label: widget.semanticLabel ?? '',
+      child: LayoutBuilder(
+        builder: (BuildContext context, BoxConstraints constraints) {
+          final double height = widget.size?.height ?? constraints.maxHeight;
+          if (_image == null || _decodedHeight != height) {
+            try {
+              _image = IconVGFile(
```

Comment:
May not be necessary for this PR, but I wonder how this will do if you have a `ListView` with lots of entries and the same `IconVG` in every entry.

That could probably be easily solved by having the `IconVG` widget take an `IconVGFile` as a parameter, and make the .asset ctor a convenience constructor that loads that from an asset repeatedly. Caching could also always be added later in a non-breaking way though.

Reply:
Ah yeah, interesting. Ideally we'd be able to plug into the ImageCache logic but that would require making that work with some abstract superclass of Image and Picture...

Reply:
Pictures are harder because there's no great way to tell their actual total size, particularly if they have `Image` objects embedded within them (wouldn't happen with this implementation but can with others, e.g. svg).